cutlery/merges.cc

```diff
@@ -0,0 +1,76 @@
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "merges.hh"
+
+Merges::Merges(std::vector<string_pair> const &merges) {
+    for (size_t i = 0; i < merges.size(); i++)
+    {
+        _merges[merges[i]] = i;
+    }
+}
+
+std::vector<std::string> Merges::apply_merges(std::vector<std::string> pieces) const {
+    // This function could be optimized more, e.g.:
+    //
+    // * Use an LRU cache for frequent inputs.
+    // * Keep a priority queue of merges rather than reconstucting
+    //   a pair list each iteration.
+    //
+    // But let's see if performance is actually an issue in practice.
+
+    while (pieces.size() != 1) {
+        auto best_pair = find_best_pair(pieces);
+        if (best_pair.first.empty() && best_pair.second.empty()) {
+            break;
+        }
+
+        std::vector<std::string> new_pieces;
+        for (size_t i = 0; i < pieces.size();) {
+            if (i < pieces.size() - 1 && pieces[i] == best_pair.first && pieces[i+1] == best_pair.second) {
+                // Merge
+                new_pieces.push_back(pieces[i] + pieces[i + 1]);
+                i += 2;
+            } else {
+                // Copy
+                new_pieces.push_back(pieces[i]);
+                ++i;
+            }
+
+        }
+
+        pieces = new_pieces;
+    }
+    
+    return pieces;
+}
+
+string_pair Merges::find_best_pair(std::vector<std::string> pieces) const {
```

Comment:
```suggestion
string_pair Merges::find_best_pair(const std::vector<std::string> &pieces) const {
```
Avoids an unnecessary copy. We could probably also rewrite this using `std::string_view` to elide even more copies.

Reply:
Ah, right. I've only ever seen `--stdlib=c++11` used in the Cython build scripts.